File: engine/fsharp_files/WordLookup.fs
namespace Scrabble.WordLookup

open System
open System.IO
open System.Text
open System.Collections.Generic
open Combinatorics

type WordLookup() = 
    static let ValidWords = new HashSet<string>()
    static let mutable OfficialWords:Map<string, string list> = Map.empty

    static do 
        File.ReadAllLines(@"..\..\..\twl.txt")
            |> Seq.map(fun w -> w.ToLower())
            |> Seq.iter(fun w -> 
                            ValidWords.Add w |> ignore
                            let alphabetized = new string (w |> Seq.sort |> Seq.toArray)

                            match OfficialWords.TryFind alphabetized with
                                // this is terrible, but you can't modify an existing member in the map
                                //pretty sure there has to be a functional way to build this map but it's evading me...
                                | Some x -> OfficialWords <- OfficialWords.Remove(alphabetized); 
                                            OfficialWords <- OfficialWords.Add(alphabetized, w :: x)
                                | None -> OfficialWords <- OfficialWords.Add(alphabetized, w::[])
                        )

    
    member this.IsValidWord (word: string): bool = 
        if String.IsNullOrEmpty word then false else ValidWords.Contains (word.ToLower())

    member this.FindAllWords (letters: char list, ?minLength: int, ?maxLength: int): string list = 
        this.Find(letters, defaultArg minLength 2, defaultArg maxLength 15)

    member this.FindWordsUsing (letters: char list, useCharAt: int, ?minLength: int, ?maxLength: int): string list = 
        this.Find(letters, defaultArg minLength 2, defaultArg maxLength 15, useCharAt)

    /// shouldn't really be a member, but you can't have optional parameters on private methods
    /// call into FindAllWords or FindwordsUsing instead - both call this
    member this.Find (letters: char list, minLength: int, maxLength: int, ?useCharAt: int): string list = 
        let useChar = defaultArg useCharAt -1
        let length = Seq.length letters
        let charAdjustedLength = match useChar with | -1 -> length | _ -> length - 1
        let max = min charAdjustedLength maxLength

        let chars = match useChar with
                        | -1 -> letters |> Seq.toArray
                        | _ -> letters // a lot of code to remove an item from a list...
                                |> List.mapi (fun index element -> (index <> useChar, element)) 
                                |> List.filter fst 
                                |> List.map snd 
                                |> Seq.toArray

        let validWords =
            [ for i in minLength .. max do
                     let generator = new CombinationGenerator(charAdjustedLength, i)
                     while generator.HasNext do
                         let indices = generator.GetNext()
                         let word = new string([| for j in 0 .. indices.Length - 1 do 
                                                    yield chars.[indices.[j]] 
                                                  match useChar with
                                                      | -1 -> ()
                                                      | _ -> yield letters.[useChar] |])
                         let possible = new string(word |> Seq.sort |> Seq.toArray)

                         match OfficialWords.TryFind(possible.ToLower()) with
                             | None -> ()
                             | Some x -> for item in x do yield item
             ]

        match useChar with
            | -1 -> validWords |> Seq.distinct |> Seq.toList
            | _ -> match OfficialWords.TryFind(letters.[useChar].ToString()) with
                    | None -> validWords |> Seq.distinct |> Seq.toList
                    | Some x -> x @ (validWords |> Seq.distinct |> Seq.toList)

File: engine/word_lookup.py
from config import GameConfig
from combination_generator import CombinationGenerator

from copy import deepcopy

class WordLookup(object):
    # def __init__(self):
    valid_words = set()
    official_words = dict()

    with open(GameConfig.wordlist_path, 'r') as f:

        for word in f.readlines():
            stripped = word.rstrip().lower()

            if len(stripped) > 1:
                valid_words.add(stripped)

                alphabetized = ''.join(sorted(stripped))

                official_words.setdefault(alphabetized, []).append(stripped)

    def is_valid_word(self, word):
        return word and (word.lower() in self.valid_words)

    def find_all_words(self, letters, min_length =2, max_length =15):
        return self.find(letters, min_length, max_length)

    def find_words_using(self, letters, use_char_at, min_length =1, max_length =15):
        return self.find(letters, min_length, max_length, use_char_at)

    def find(self, ltrs, min_length, max_length, use_char_at =-1):
        length = len(ltrs)
        letters = [deepcopy(l) for l in ltrs]


        if use_char_at == -1:
            use_char = use_char_at
            char_adjusted_length = length
            chars = letters
        else:
            use_char = use_char_at
            char_adjusted_length = length - 1
            chars = [letters[i] for i in range(0, len(letters)) if i != use_char]

        maxlength = min(char_adjusted_length, max_length)
        validwords = set()

        for i in range(min_length, maxlength+1):
            generator = CombinationGenerator(char_adjusted_length, i)

            while generator.has_next():
                indices = generator.get_next()

                word = []

                for j in range(0, len(indices)):
                    word.append(chars[indices[j]])
                
                if use_char != -1:
                    word.append(letters[use_char])

                word.sort()
                possible = ''.join(word)

                results = deepcopy(self.official_words.get(possible.lower()))

                if results:
                    validwords.update(results)

        if use_char == -1:
            return list(validwords)
        else:
            x = deepcopy(self.official_words.get(letters[use_char]))
            if x:
               return x + list(validwords)
            else:
                return list(validwords)
